File: solutions/2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func nodesBetweenCriticalPoints(head *ListNode) []int {
	first := -1

	last := -1

	minDistance := int(^uint(0) >> 1)

	position := 1

	prev := head

	curr := head.Next

	for curr != nil && curr.Next != nil {
		isCritical := (curr.Val > prev.Val && curr.Val > curr.Next.Val) ||
			(curr.Val < prev.Val && curr.Val < curr.Next.Val)

		if isCritical {
			if first == -1 {
				first = position
			} else {
				distance := position - last

				if distance < minDistance {
					minDistance = distance
				}
			}

			last = position
		}

		prev = curr

		curr = curr.Next

		position++
	}

	if first == -1 || first == last {
		return []int{-1, -1}
	}

	maxDistance := last - first

	return []int{minDistance, maxDistance}
}

File: solutions/2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* nodesBetweenCriticalPoints(struct ListNode* head, int* returnSize) {
    int capacity = 100000;
    int* nums = (int*)malloc(capacity * sizeof(int));
    int n = 0;

    while (head) {
        nums[n++] = head->val;
        head = head->next;
    }

    int* points = (int*)malloc(n * sizeof(int));
    int m = 0;

    for (int i = 1; i < n - 1; i++) {
        if (
            (nums[i] > nums[i - 1] && nums[i] > nums[i + 1]) ||
            (nums[i] < nums[i - 1] && nums[i] < nums[i + 1])
        ) {
            points[m++] = i;
        }
    }

    int* result = (int*)malloc(2 * sizeof(int));
    *returnSize = 2;

    if (m < 2) {
        result[0] = -1;
        result[1] = -1;

        free(nums);
        free(points);

        return result;
    }

    int minDist = INT_MAX;
    int maxDist = points[m - 1] - points[0];

    for (int i = 1; i < m; i++) {
        int dist = points[i] - points[i - 1];

        if (dist < minDist) {
            minDist = dist;
        }
    }

    result[0] = minDist;
    result[1] = maxDist;

    free(nums);
    free(points);

    return result;
}